### DISCUSSION TITLE
Composing environment variables

### DISCUSSION THREAD

**Discussion by tyleralbee** [vercel/turborepo#9458] [Help] [answered]
### Summary

I am trying to wrap my head around how environment variables should be handled. I understand that instead of having a single `.env` file in the root of the repository, [`.env` files should be kept in the packages where they're used](https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#use-env-files-in-packages):

> Using a `.env` file at the root of the repository is not recommended. Instead, we recommend placing your `.env` files into the packages where they're used.

First: is the above usage of "packages" is referring to [Application Packages](https://turbo.build/repo/docs/core-concepts/package-types#application-packages), [Internal Packages](https://turbo.build/repo/docs/core-concepts/internal-packages), or both?

The [surrounding context](https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#use-env-files-in-packages) leads me to believe that "packages" is referring to [Application Packages](https://turbo.build/repo/docs/core-concepts/package-types#application-packages):

> This practice more closely models the runtime behavior of your **applications** since environment variables exist in each **application**'s runtime individually.

It makes sense to me that the environment variables would be application-scoped, but please correct me if I'm wrong. To be clear, my understanding is that the repository should look like this:

```text
my-repository
├── apps
│   └── app-a
│   │   └── .env
│   │   └── .env.development
│   │   └── ... ... ...
│   │   └── package.json
│   └── app-b
│   │   └── .env
│   │   └── .env.development
│   │   └── ... ... ...
│   │   └── package.json
└── packages
│   └── package-a
│   └── package-b
└── package.json
└── README.md
```

In the structure above, `app-a` and `app-b` have `.env` files, whereas `package-a`, `package-b`, and the repository root do not.

In some cases, I feel it would make sense for my `.env` files to be co-located with an [Internal Packages](https://turbo.build/repo/docs/core-concepts/internal-packages). For example, I have a `specific-db` [Internal Package](https://turbo.build/repo/docs/core-concepts/internal-packages) which is always looking for `SPECIFIC_DB_CONNECTION_URL`. This variable is dependent on the environment; that is, `app-a` and `app-b` will pass the same `SPECIFIC_DB_CONNECTION_URL` in production. In addition, `specific-db` has code generation scripts that rely on `SPECIFIC_DB_CONNECTION_URL`.

If I structure my repository such that my `.env` files are co-located with `specific-db`:

```text
my-repository
├── apps
│   └── app-a
│   └── app-b
└── packages
│   └── specific-db
│   |   └── scripts
│   |   |   └── codegen.ts
│   │   └── .env
│   │   └── .env.development
│   │   └── ... ... ...
│   │   └── package.json
│   └── package-b
└── package.json
└── README.md
```

How would I go about composing the environment variables for `app-a` from the environment variables in `specific-db`? Is there a better way than duplicating the `.env` files from `specific-db` and/or writing codegen to keep them in sync?

Unrelated question: what's a use-case in which `globalEnv` makes sense to use over task-specific `env`? Can someone give me an example of an environment variable that should result in a `lint` task missing cache on change?

### References

- [**Using environment variables**](https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables)
- [**System environment variables**](https://turbo.build/repo/docs/reference/system-environment-variables)
- [**Deployment using the wrong environment variables**](https://turbo.build/repo/docs/crafting-your-repository/constructing-ci#deployment-using-the-wrong-environment-variables)

**Comment by anthonyshew** [comment, marked as answer]
Good question, and you're thinking about things attentively. I appreciate the thoughtfully worded question(s).

The way that I like thinking about it is: When I run my code locally, I'm simulating what happens in production. You don't run packages, you run applications that are composed of packages (which is something you correctly alluded to).

With that mental model, you start to realize that _a package has no independent runtime_. It _only_ runs within the context of an application. So, if I want to model that locally, that would mean I only use `.env` files in applications.

There's an exception: If you're running scripts directly in packages that require environment variables, you might be forced to put a `.env` file into a package. [Prisma](https://turbo.build/repo/docs/guides/tools/prisma#run-prisma-init) comes to mind. This is indeed a little awkward and breaks the idealistic model I was describing earlier, but is not impossible to do and create great workflows with.

**Reply by tyleralbee** [reply]
Thanks, that makes sense.

**Comment by raky291** [comment]
Hi everyone ✌️ 

Do you know where can I find more resources to learn how to configure .env files in Turborepo? I still don't understand how I should configure them (with turbo.json, root vs packages, etc.). My structure is similar, app-1, app-2, package-3 and they all need to use .env files.

Are there any videos, or example in the repository, or something like that?

**Reply by tyleralbee** [reply]
I'll share where I've ended up. I'm going to break it down into two parts:

1. [Foundation](#foundation): What Turborepo is expecting your project to look like
2. [Extension](#extension): How I have built upon this foundation in my own projects

## Foundation

Here's some tenets I try to stick to:

- Environment variables (`.env` files) are defined in the `apps/*` that need them. They are **never** defined in the workspace root, and are **rarely** defined in `packages/*`.
- We should **avoid** adding environment variables to `globalEnv` in our root `turbo.json`.
- Our root `turbo.json`'s `globalPassThroughEnv` should only contain environment variables that _we did not define ourselves_.
- Our root `turbo.json`'s `tasks` should **not** define `env`; this should always be delegated to the `turbo.json` files in our packages.
- Our root `turbo.json`'s `tasks.build` **should** add `.env.*` and `.env` to its `inputs`.
- Our root `turbo.json`'s `globalDependencies` **should** include `.env` files that are not tracked by `git`.
- Packages whose source code reference an environment variable directly **should** include the environment variable in `tasks.build.env`.
- Apps that reference an environment variable indirectly through a package should **not** include the environment variable in `tasks.build.env` - only the package should.

In summary:

- `/turbo.json:globalEnv` should be `[]`
- `/turbo.json:globalPassThroughEnv` includes [environment variables set in CI](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables), 3rd party environment variables (e.g. [npm's `package.json` vars](https://docs.npmjs.com/cli/v8/using-npm/scripts#packagejson-vars)) and any [system environment variables populated by your cloud platform](https://vercel.com/docs/projects/environment-variables/system-environment-variables).
- `/turbo.json:globalDependencies` includes the `.env` files that are not tracked by `git` (e.g. `[... ".env.*", ".env"]`)
- `/turbo.json:tasks.build.env` should not be set
- `/turbo.json:tasks.build.inputs` should include your environment variable files (e.g. `["$TURBO_DEFAULT$", ... ".env.*", ".env"]`)
- `/packages/example/turbo.json:tasks.build.env` should include environment variables used by the package
- `/apps/something/turbo.json:tasks.build.env` should **not** include environment variables used by its dependencies (unless the app references the environment variable itself)

This section is meant to be unopinionated. If anything I've written here goes against best practice, please let me know.

## Extension

I personally follow [Vite's method](https://vite.dev/guide/env-and-mode#env-files) of structuring my `.env` files:

```sh
.env                # loaded in all cases
.env.local          # loaded in all cases, ignored by git
.env.[mode]         # only loaded in specified mode
.env.[mode].local   # only loaded in specified mode, ignored by git
```

The `turbo.json` files in my `apps/` are never much more than this:

```json
{
    "$schema": "https://turbo.build/schema.json",
    "extends": ["//"],
    "tasks": {
        "build": {
            "env": ["PREFIX_USED_BY_APP_*"]
        },
        "dev": {
            "env": ["PREFIX_USED_BY_APP_*"]
        }
    }
}
```

The `turbo.json` files in my `packages/` are usually limited to this:

```json
{
    "$schema": "https://turbo.build/schema.json",
    "extends": ["//"],
    "tasks": {
        "build": {
            "env": ["PREFIX_USED_BY_PACKAGE_*"]
        }
    }
}
```

**Reply by raky291** [reply]
@tyleralbee Thank you so much, this helped me a lot.

**Reply by nktnet1** [reply]
@tyleralbee thanks for the insight - this was extremely helpful!

For anyone looking for a concrete example, I've implemented the suggestions above in my boilerplate here:
- https://github.com/nktnet1/rt-stack

Initially I add a separate package called `@repo/env` that gets installed in other apps/packages - was able to convert almost all the packages to be pure (they now take input args to instantiate - factory pattern), so env is no longer used.

One exception is the `@repo/db`, which if you use ORMs like prisma/drizzle, you'll still need to still have a DATABASE_URL env in order to run migrations for those packages.

**Comment by kelvinsekx** [comment]
None of this has tend to work for me.
Here is how I have been doing things:


```bash
Root > apps
             > website
       > packages
             > http

```
The Root turbo.json has

```bash
{
  "$schema": "https://turbo.build/schema.json",
  "ui": "tui",
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "inputs": ["$TURBO_DEFAULT$", ".env*"],
      "outputs": [".next/**", "!.next/cache/**", "dist/**"]
    },
    "lint": {
      "dependsOn": ["^lint"]
    },
    "dev": {
      "dependsOn": ["^dev"],
      "cache": false,
      "persistent": false
    }
  },
  "globalEnv": []
}

```

The @repo/http has

```bash
{
  "$schema": "https://turborepo.org/schema.json",
  "extends": ["//"],
  "tasks": {
    "dev": {
      "env": ["API_*"],
      "inputs": ["$TURBO_DEFAULT$", ".env"]
    },
    "build": {
      "env": ["API_*"],
      "inputs": ["$TURBO_DEFAULT$", ".env"]
    }
  }
}

```

The code where I am trying to use the .env variables

```bash
import axios, { AxiosRequestConfig, AxiosInstance } from "axios";
import path from "path";
import dotenv from "dotenv";

dotenv.config({
  path: path.resolve(process.cwd(), ".env"),
});


// Get Base URL
const BASE_URL = process.env.API_BASE_URL;

const BASE_CONFIG: AxiosRequestConfig = {
  baseURL: BASE_URL,
  timeout: 3000,
  headers: {
    "Content-Type": "application/json",
  },
};

const baseClient: AxiosInstance = axios.create(BASE_CONFIG);

export default baseClient;

```

**Reply by kelvinsekx** [reply]
> None of this has tend to work for me. Here is how I have been doing things:
> 
> ```shell
> Root > apps
>              > website
>        > packages
>              > http
> ```
> 
> The Root turbo.json has
> 
> ```shell
> {
>   "$schema": "https://turbo.build/schema.json",
>   "ui": "tui",
>   "tasks": {
>     "build": {
>       "dependsOn": ["^build"],
>       "inputs": ["$TURBO_DEFAULT$", ".env*"],
>       "outputs": [".next/**", "!.next/cache/**", "dist/**"]
>     },
>     "lint": {
>       "dependsOn": ["^lint"]
>     },
>     "dev": {
>       "dependsOn": ["^dev"],
>       "cache": false,
>       "persistent": false
>     }
>   },
>   "globalEnv": []
> }
> ```
> 
> The @repo/http has
> 
> ```shell
> {
>   "$schema": "https://turborepo.org/schema.json",
>   "extends": ["//"],
>   "tasks": {
>     "dev": {
>       "env": ["API_*"],
>       "inputs": ["$TURBO_DEFAULT$", ".env"]
>     },
>     "build": {
>       "env": ["API_*"],
>       "inputs": ["$TURBO_DEFAULT$", ".env"]
>     }
>   }
> }
> ```
> 
> The code where I am trying to use the .env variables
> 
> ```shell
> import axios, { AxiosRequestConfig, AxiosInstance } from "axios";
> import path from "path";
> import dotenv from "dotenv";
> 
> dotenv.config({
>   path: path.resolve(process.cwd(), ".env"),
> });
> 
> 
> // Get Base URL
> const BASE_URL = process.env.API_BASE_URL;
> 
> const BASE_CONFIG: AxiosRequestConfig = {
>   baseURL: BASE_URL,
>   timeout: 3000,
>   headers: {
>     "Content-Type": "application/json",
>   },
> };
> 
> const baseClient: AxiosInstance = axios.create(BASE_CONFIG);
> 
> export default baseClient;
> ```

I unblocked myself after a few tweaks. The following below are my explanations:

1. Internal packages (in /packages) should NEVER use (or have) its .env file.
2. Interpret internal packages as a module used by apps.
3. Since internal packages are just modules consumed by apps (in apps/), the behavior of internal packages is entirely determined by the /apps and its environment.
4. .env files should only reside (or LIVE) in /apps and be called in internal packages like they existed in the /apps. for my example above, since the app/website is a NEXT app, I would put the .env to be referenced in packages/http

Let's take previous code for example

```bash
// packages/http

import axios, { AxiosRequestConfig, AxiosInstance } from "axios";


// Get Base URL
const BASE_URL = process.env.NEXT_PUBLIC__API_BASE_URL; // use it like it would be typically used in next app because the package that consumes it is a NEXT app

const BASE_CONFIG: AxiosRequestConfig = {
  baseURL: BASE_URL,
  timeout: 3000,
  headers: {
    "Content-Type": "application/json",
  },
};

const baseClient: AxiosInstance = axios.create(BASE_CONFIG);

export default baseClient;

```

```bash
// apps/website/.env

NEXT_PUBLIC_API_BASE_URL=https://<domain>
```

cc @fa-901. Thank you for reaching out.

**Comment by arataqueh** [comment]
Has anyone had issues with NODE_EXTRA_CA_CERTS global var which is usually set on .zhsrc or .bashrc files and this works fine with my pnpm dev command directly run on apps/app folder but somehow with turbo dev command this global var is not captured. Im using mac and vscode by the way so I appreciate if anyone has found a way to make turbo dev command work with global vars without having to specify them on .env file and on turbo.json task like below:

"dev": {
      "env": [
        "NODE_EXTRA_CA_CERTS"
      ],
      "cache": false,
      "persistent": true
    }

**Reply by tyleralbee** [reply]
> Has anyone had issues with NODE_EXTRA_CA_CERTS global var which is usually set on .zhsrc or .bashrc files and this works fine with my pnpm dev command directly run on apps/app folder but somehow with turbo dev command this global var is not captured. Im using mac and vscode by the way so I appreciate if anyone has found a way to make turbo dev command work with global vars without having to specify them on .env file and on turbo.json task like below:
> 
> "dev": { "env": [ "NODE_EXTRA_CA_CERTS" ], "cache": false, "persistent": true }

If you haven't already, make sure the environment variable is there in [Loose Mode](https://turborepo.com/docs/crafting-your-repository/using-environment-variables#loose-mode). 

Just to be clear - you're aware you _need_ to specify environment variables for them to be passed through, correct?

**Reply by arataqueh** [reply]
I am aware but I don't know if you understood my question. This NODE_EXTRA_CA_CERTS variable is to set the path for Certificate Authority (CA) file and this is to avoid those UNABLE_TO_GET_ISSUER_CERT_LOCALLY issues since I also work on a corporate environment. This is one of those variables you set on .zhsrc or .bashrc file of your local machine. When I run a pnpm/npm dev directly on the app folder, it works fine and it reads this global var from my .zhsrc file but somehow turbo does not and I assume this is because turbo creates a subshell when it run its tasks?

If I specify the NODE_EXTRA_CA_CERTS var in my .env file and also add it to the list of env vars within the env prop of the dev script within root package.json like I showed previously, it works but this is not ideal because as I said it works fine with pnpm and npm commands which access the .zhsrc or .bashrc file for local development. Unless this is a limitation of turbo and it cannot read from those files in your machine.

**Reply by anthonyshew** [reply]
There is no limitation that matches what you're describing. It is most likely that @tyleralbee is pointing in the right direction.
